### PR TITLE
Integrate new routing functions

### LINKED ISSUE #14450: Integrate New Route Azure Functions
### User Story

As ReportStream,
the route step shall be split up into two separate steps per the [SRD](https://cdc.sharepoint.com/:w:/r/teams/ReportStream/Shared%20Documents/Engineering/Design/SRD/SRD-Routing.docx?d=w3a2fa783e8dc4e9b8397da9f5d968dfa&csf=1&web=1&e=9wJcJi)

### Description/Use Case

- The Route function shall be replaced by the [Destination Filter Function](https://github.com/CDCgov/prime-reportstream/issues/13691) and [Implement Receiver Filter Function](https://github.com/CDCgov/prime-reportstream/issues/13692) to achieve the benefits outlined in the [SRD](https://cdc.sharepoint.com/:w:/r/teams/ReportStream/Shared%20Documents/Engineering/Design/SRD/SRD-Routing.docx?d=w3a2fa783e8dc4e9b8397da9f5d968dfa&csf=1&web=1&e=9wJcJi)
- By the time this ticket is picked up, both new functions shall be implemented but not integrated into the application (the convert step shall still call the old Route function. To complete this ticket, we must perform the following operations in order:

  1. update the convert step to route to the destination filter function instead of route function
  2. deploy to staging and master
  3. verify there are no queue messages for the route step in staging and master
  4. update the pipeline to delete the Route step code
  5. deploy to staging and master

### Risks/Impacts/Considerations

We will need to verify there are no tools, queries, etc. depending on the current Route step existing. Lineage tracking should just work, but there may be other things that may depend on this step, like any queries the engagement team is using that directly query the Route step.
- Look through Metabase queries in both staging and prod 
- Review findings with Engagement team 
- look through queries in repo

If there are existing functionalities depending on the Route step, they should be updated as part of this ticket.

### Acceptance Criteria

- [ ] Route function replaced with the new functions and marked for removal
- [ ] Staging and Prod environments working with new functions
- [ ] Any queries or dependencies on the Route step have been updated
- [ ] Route Step documentation updated or follow-up ticket created
- [ ] ReportRouteEvent.originalObservations is covered in unit and/or integration tests
- [ ] ReportRouteEvent.receiver is no longer nullable (moved to cleanup ticket)

### PR DESCRIPTION
**Pull request by thetaurean**
This PR integrates the new routing functions, `DestinationFilter` and `ReceiverFilter`, and phases out dependencies on FHIRRouter. It also relocated/refactored several test utility functions associated with creating receivers/orgs as well as reports and their lineages.

Test Steps:
1. Run test suites

## Changes
- Now utilizing new routing functions
  - Convert step now routes to DestinationFilter step
  - Which then goes ReceiverFilter->Translate to resume pipeline
- `FhirRouterIntegrationTests` split into `FhirDestinationFilterIntegrationTests` and `FhirReceiverFilterIntegrationTests` 
  - Common utilities reworked & moved to `UniversalPipelineTestUtils.kt`
  - All assertions reworked to improve output
  - Added 4 observation filter test cases to `FhirReceiverFilterIntegrationTests` (1 condition + 3 mapped condition)

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [x] Added tests?

## Linked Issues
- Fixes #14450 

## To Be Done
- #15039

### CODE REVIEW

File: prime-router/docs/standard-operating-procedures/configuring-filters.md
Comment (thetaurean) [inline prime-router/docs/standard-operating-procedures/configuring-filters.md:1]: This is all existing content.

**Review by jack-h-wang — APPROVED** [review]
LGTM, tests passing locally

File: prime-router/docs/universal-pipeline/receiver-filter.md [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+catch(e: SchemaException) {
+    actionLogger?.warn(EvaluateFilterConditionErrorMessage(e.message))
+    exceptionFilters += filterElement
+}
+```
+
+Various scenarios for filter logging are outlined in the [filtering design](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/design/features/0001-universal-pipeline-filter-reporting.md#filter-log-scenarios).
+
+## Events
+
```
Comment (arnejduranovic) [inline prime-router/docs/universal-pipeline/receiver-filter.md:317]: Please complete?

File: prime-router/docs/universal-pipeline/destination-filter.md [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    actionLogger?.warn(EvaluateFilterConditionErrorMessage(e.message))
+    exceptionFilters += filterElement
+}
+```
+
+Various scenarios for filter logging are outlined in
+the [filtering design](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/design/features/0001-universal-pipeline-filter-reporting.md#filter-log-scenarios).
+
+## Events
+
```
[the remark this answers is not in the document]
Comment (thetaurean) [reply prime-router/docs/universal-pipeline/destination-filter.md:210]: I added it 😝 and forgot to fill it in

File: prime-router/docs/standard-operating-procedures/configuring-filters.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            organizationName: yoyodyne
+            topic: full-elr
+            customerStatus: active
+            jurisdictionalFilter: [ "(Bundle.entry.resource.ofType(ServiceRequest)[0].requester.resolve().organization.resolve().address.state.exists() and Bundle.entry.resource.ofType(ServiceRequest)[0].requester.resolve().organization.resolve().address.state = 'CA') ]
+```
+
+2. Use the following commands to load the information from the .yml files into the staging database. First obtain a
+   login token for staging
+
+`./prime login –env staging`
```
Comment (adegolier) [inline prime-router/docs/standard-operating-procedures/configuring-filters.md:32]:
```suggestion
`./prime login --env staging`
```

File: prime-router/docs/standard-operating-procedures/configuring-filters.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+Next update the staging DB
+
+`./prime multiple-settings set –env staging –input <file-location>`
```
Comment (adegolier) [inline prime-router/docs/standard-operating-procedures/configuring-filters.md:36]:
```suggestion
`./prime multiple-settings set --env staging --input <file-location>`
```

File: prime-router/docs/standard-operating-procedures/configuring-filters.md (hunk 3 above)
Comment (adegolier) [reply prime-router/docs/standard-operating-procedures/configuring-filters.md:32]: Guessing you copy/pasted these out of the deployment docs. Word likes to merge the double dashes into an em dash, which the command will get mad about.

File: prime-router/src/main/kotlin/fhirengine/azure/FHIRFunctions.kt [hunk 5]
```diff
@@ -65,12 +65,13 @@ class FHIRFunctions(
...
         messagesToDispatch.forEach {
             queueAccess.sendMessage(
-                elrRoutingQueueName,
+                elrDestinationFilterQueueName,
                 it.serialize()
             )
         }
     }
 
+    // TODO: deprecated (see https://github.com/CDCgov/prime-reportstream/issues/15039)
```
Comment (adegolier) [inline prime-router/src/main/kotlin/fhirengine/azure/FHIRFunctions.kt:74]: Is there a reason you're deprecating with TODOs and not with the Deprecated annotation?

File: prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt [hunk 6]
```diff
@@ -129,9 +129,7 @@ fun Bundle.getObservations() = this.entry.map { it.resource }.filterIsInstance<O
 
 fun Bundle.getObservationsWithCondition(codes: List<String>): List<Observation> =
     if (codes.isEmpty()) {
-        // TODO: consider throwing IllegalArgumentException here while implementing
-        //  https://github.com/CDCgov/prime-reportstream/issues/12705
-        emptyList()
+        throw IllegalStateException("Invalid mapped condition filter")
```
Comment (adegolier) [inline prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt:132]: InvalidState is meant to be used when a method is invoked at a time when the environment or app is not in a correct state for the call.... IllegalArgument seems more appropriate here.

Comment (thetaurean) [reply prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt:132]: Makes more sense to me 👍  (and past me looking at the comment xD)

File: prime-router/src/main/kotlin/fhirengine/azure/FHIRFunctions.kt (hunk 5 above)
Comment (thetaurean) [reply prime-router/src/main/kotlin/fhirengine/azure/FHIRFunctions.kt:74]: Historically that's how we've handled code that is pending removal/cleanup that isn't externally consumed. I reworded the comment to clean-up for now and updated the referencing ticket

I'm not 100% I should deprecate with an annotation but happy to do so if we feel strongly.

File: prime-router/docs/standard-operating-procedures/configuring-filters.md (hunk 3 above)
Comment (thetaurean) [reply prime-router/docs/standard-operating-procedures/configuring-filters.md:32]: Nice catch 🎉 🥇! I took them straight out of another markdown and neglected to proofread.

**Review by arnejduranovic — APPROVED** [review]

**Pull request merged**